clippy_lints/src/uninit_vec.rs

```diff
@@ -0,0 +1,183 @@
+fn extract_with_capacity_or_reserve_target(cx: &LateContext<'_>, stmt: &'tcx Stmt<'_>) -> Option<LocalOrExpr<'tcx>> {
+    match stmt.kind {
+        StmtKind::Local(local) => {
+            // let mut x = Vec::with_capacity()
+            if_chain! {
+                if let Some(init_expr) = local.init;
+                if let PatKind::Binding(_, hir_id, _, None) = local.pat.kind;
+                if is_with_capacity(cx, init_expr);
+                then {
+                    Some(LocalOrExpr::Local(hir_id))
+                } else {
+                    None
+                }
+            }
+        },
+        StmtKind::Expr(expr) | StmtKind::Semi(expr) => {
+            match expr.kind {
+                ExprKind::Assign(lhs, rhs, _span) if is_with_capacity(cx, rhs) => {
+                    // self.vec = Vec::with_capacity()
+                    Some(LocalOrExpr::Expr(lhs))
+                },
+                ExprKind::MethodCall(path, _, [self_expr, _], _) => {
+                    // self.vec.reserve()
+                    if is_type_diagnostic_item(cx, cx.typeck_results().expr_ty(self_expr).peel_refs(), sym::vec_type)
+                        && path.ident.name.as_str() == "reserve"
+                    {
+                        Some(LocalOrExpr::Expr(self_expr))
+                    } else {
+                        None
+                    }
+                },
+                _ => None,
+            }
+        },
+        StmtKind::Item(_) => None,
+    }
+}
+
+fn is_with_capacity(cx: &LateContext<'_>, expr: &'tcx Expr<'_>) -> bool {
+    if_chain! {
+        if let ExprKind::Call(path_expr, _) = &expr.kind;
+        if let ExprKind::Path(qpath) = &path_expr.kind;
+        if let Res::Def(_, def_id) = cx.qpath_res(qpath, path_expr.hir_id);
+        then {
+            match_def_path(cx, def_id, &paths::VEC_WITH_CAPACITY)
+        } else {
+            false
+        }
+    }
+}
+
+/// Returns self if the expression is `Vec::set_len()`
+fn extract_set_len_self(cx: &LateContext<'_>, expr: &'tcx Expr<'_>) -> Option<(&'tcx Expr<'tcx>, Span)> {
+    // peel unsafe blocks in `unsafe { vec.set_len() }`
+    let expr = peel_hir_expr_while(expr, |e| {
+        if let ExprKind::Block(block, _) = e.kind {
+            // Extract the first statement/expression
+            match (block.stmts.get(0).map(|stmt| &stmt.kind), block.expr) {
+                (None, Some(expr)) => Some(expr),
+                (Some(StmtKind::Expr(expr) | StmtKind::Semi(expr)), _) => Some(expr),
+                _ => None,
+            }
+        } else {
+            None
+        }
+    });
+    match expr.kind {
+        ExprKind::MethodCall(_, _, [vec_expr, _], _) => {
+            cx.typeck_results().type_dependent_def_id(expr.hir_id).and_then(|id| {
+                if match_def_path(cx, id, &paths::VEC_SET_LEN) {
+                    Some((vec_expr, expr.span))
+                } else {
+                    None
+                }
+            })
```

Comment:
Here it would also be better to use diagnostic items instead. The code would look like this
```rs
        ExprKind::MethodCall(path, _, [self_expr, _], _) => {
            let self_type = cx.typeck_results().expr_ty(self_expr).peel_refs();
            if is_type_diagnostic_item(cx, self_type, sym::vec_type)
                && path.ident.name.as_str() == "set_len"
            {
                Some((self_expr, expr.span))
            } else {
                None
            }
```